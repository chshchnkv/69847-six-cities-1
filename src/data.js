export const AccommodationType = {
  APARTMENT: `apartment`,
  ROOM: `room`,
  HOUSE: `house`,
  HOTEL: `hotel`,
  HOSTEL: `hostel`,
};

export const PeriodType = {
  NIGHT: `night`
};

export const urlSortField = `sortby`;
export const urlSortOrder = `order`;

export const SortField = {
  ID: `id`,
  PRICE: `price`,
  RANK: `rank`
};

export const SortOrder = {
  ASC: `asc`,
  DESC: `desc`
};

export const sortOptions = [
  {
    title: `Popular`,
    sort: {
      field: SortField.ID,
      order: SortOrder.ASC
    }
  },
  {
    title: `Price: low to high`,
    sort: {
      field: SortField.PRICE,
      order: SortOrder.ASC
    }
  },
  {
    title: `Price: high to low`,
    sort: {
      field: SortField.PRICE,
      order: SortOrder.DESC
    }
  },
  {
    title: `Top rated first`,
    sort: {
      field: SortField.RANK,
      order: SortOrder.DESC
    }
  }
];

export const ratings = [
  {
    value: 1,
    title: `terribly`
  },
  {
    value: 2,
    title: `badly`
  },
  {
    value: 3,
    title: `not bad`
  },
  {
    value: 4,
    title: `good`
  },
  {
    value: 5,
    title: `perfect`
  },
];

export const MAX_RATING_IN_STARS = 5;
