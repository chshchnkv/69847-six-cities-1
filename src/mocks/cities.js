export const Cities = [
  {
    id: 1,
    title: `Amsterdam`,
    longitude: 52.38333,
    latitude: 4.9,
    zoom: 12,
  },
  {
    id: 2,
    title: `Paris`,
    longitude: 48.8589507,
    latitude: 2.2770204,
    zoom: 12,
  },
  {
    id: 3,
    title: `Cologne`,
    longitude: 50.9578353,
    latitude: 6.8272405,
    zoom: 12,
  },
  {
    id: 4,
    title: `Brussels`,
    longitude: 50.8550625,
    latitude: 4.3053506,
    zoom: 12,
  },
  {
    id: 5,
    title: `Hamburg`,
    longitude: 53.5586941,
    latitude: 9.78774,
    zoom: 12,
  },
  {
    id: 6,
    title: `Dusseldorf`,
    longitude: 51.2385861,
    latitude: 6.6742684,
    zoom: 12,
  },
];
