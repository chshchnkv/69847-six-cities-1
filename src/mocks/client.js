import {JSDOM} from "jsdom";

const {document} = (new JSDOM(`<body></body>`)).window;
global.document = document;

const mapElement = global.document.createElement(`div`);
mapElement.setAttribute(`id`, `map`);
global.document.body.appendChild(mapElement);
