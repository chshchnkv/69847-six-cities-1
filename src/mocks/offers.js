import {AccommodationType, PeriodType} from "../data";

export const offers = [
  {
    id: 1,
    title: `Beautiful & luxurious apartment at great location`,
    src: `img/apartment-01.jpg`,
    isPremium: false,
    price: {
      value: 120,
      period: PeriodType.NIGHT
    },
    type: AccommodationType.APARTMENT,
    rating: 3.2,
  }, {
    id: 2,
    title: `Wood and stone place`,
    src: `img/apartment-02.jpg`,
    isPremium: false,
    price: {
      value: 60,
      period: PeriodType.NIGHT
    },
    type: AccommodationType.HOUSE,
    rating: 2.8,
  }, {
    id: 3,
    title: `Canal View Prinsengracht`,
    src: `img/apartment-03.jpg`,
    isPremium: false,
    price: {
      value: 30,
      period: PeriodType.NIGHT
    },
    type: AccommodationType.HOTEL,
    rating: 4.2,
  }, {
    id: 4,
    title: `Nice, cozy, warm big bed apartment`,
    src: `img/apartment-01.jpg`,
    isPremium: true,
    price: {
      value: 300
    },
    type: AccommodationType.APARTMENT,
    rating: 5,
  }, {
    id: 5,
    title: `ibis Styles`,
    src: `img/studio-01.jpg`,
    isPremium: false,
    price: {
      value: 110,
      period: PeriodType.NIGHT
    },
    type: AccommodationType.HOTEL,
    rating: 1,
  },
];
