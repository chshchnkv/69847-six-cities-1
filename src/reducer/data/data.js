import {getCityInfoByName, getNewId, getOfferIndexById, sortOffers} from "../../utils";
import {SortField, SortOrder} from "../../data";

const serverOfferToOffer = (serverOffer, cities) => {
  const {
    city,
    ...offer
  } = serverOffer;

  const {
    location,
    ...offerWithNoLocation
  } = offer;

  const {name: cityName} = city;

  const cityInfo = getCityInfoByName(cities, cityName);
  const {id: cityId} = cityInfo;

  return {
    ...offerWithNoLocation,
    ...{city: cityId},
    ...{location: {
      longitude: location.latitude,
      latitude: location.longitude,
    }}
  };
};

const initialState = {
  cities: [],
  offers: [],
  favorites: [],
  reviews: [],
  sortOptions: {
    field: SortField.ID,
    order: SortOrder.ASC
  },
};

export const Action = {
  SET_CITIES: `load_cities`,
  SET_OFFERS: `set_offers`,
  SET_FAVORITES: `set_favorites`,
  SET_REVIEWS: `load_reviews`,
  SORT_OPTIONS: `sort`,
};

export const ActionCreator = {
  [Action.SET_OFFERS]: (offers) => ({
    type: Action.SET_OFFERS,
    payload: offers
  }),

  [Action.SET_CITIES]: (cities) => ({
    type: Action.SET_CITIES,
    payload: cities
  }),

  [Action.SET_REVIEWS]: (reviews) => ({
    type: Action.SET_REVIEWS,
    payload: reviews
  }),

  [Action.SET_FAVORITES]: (favorites) => ({
    type: Action.SET_FAVORITES,
    payload: favorites
  }),

  [Action.SORT_OPTIONS]: (sortOptions) => ({
    type: Action.SORT_OPTIONS,
    payload: sortOptions
  }),

  [Action.SET_COMMENT_SENDING]: (isSendingComment) => ({
    type: Action.SET_COMMENT_SENDING,
    payload: isSendingComment
  }),
};

export const Operation = {
  loadOffers: () => (dispatch, _getState, api) => {
    return api.get(`/hotels`)
      .then((response) => {
        const citiesMap = new Map();
        const offers = [];
        const serverOffers = response.data;
        for (let serverOffer of serverOffers) {
          const {
            city,
            ...offer
          } = serverOffer;

          const {
            location,
            ...offerWithNoLocation
          } = offer;

          const {name: cityName} = city;

          let cityId;
          if (citiesMap.has(cityName)) {
            cityId = citiesMap.get(cityName).id;
          } else {
            cityId = getNewId();
            citiesMap.set(cityName, {...city, ...{id: cityId}});
          }

          offers.push({
            ...offerWithNoLocation,
            ...{city: cityId},
            ...{location: {
              longitude: location.latitude,
              latitude: location.longitude,
            }
            }
          });
        }

        const cities = [...citiesMap.values()];

        dispatch(ActionCreator[Action.SET_CITIES](cities));
        dispatch(ActionCreator[Action.SET_OFFERS](offers));

        return cities.length > 0 ? cities[0].id : -1;
      });
  },

  loadComments: (propertyId) => (dispatch, _getState, api) => {
    return api.get(`/comments/${propertyId}`)
      .then((response) => {
        dispatch(ActionCreator[Action.SET_REVIEWS](response.data));
      });
  },

  loadFavorites: () => (dispatch, _getState, api) => {
    return api.get(`/favorite`)
      .then((response) => {
        const {data: {cities}} = _getState();
        const sortedCities = cities.slice(0).sort((a, b) => {
          const {name: nameA} = a;
          const {name: nameB} = b;
          const cityNameA = nameA.toLowerCase().trim();
          const cityNameB = nameB.toLowerCase().trim();
          if (cityNameA === cityNameB) {
            return 0;
          } else if (cityNameA > cityNameB) {
            return 1;
          } else {
            return -1;
          }
        });

        const favoriteOffers = response.data.map((offer) => serverOfferToOffer(offer, cities));

        const result = sortedCities
          .map((city) => ({
            cityId: city.id,
            offers: favoriteOffers.filter((offer) => offer.city === city.id)
          }))
          .filter((item) => item.offers.length > 0);

        dispatch(ActionCreator[Action.SET_FAVORITES](result));
      })
      .catch(() => alert(`Something went wrong :(`));
  },


  postComment: (propertyId, rating, comment) => (dispatch, _getState, api) => {
    dispatch(ActionCreator[Action.SET_COMMENT_SENDING](true));
    return api.post(`/comments/${propertyId}`, {
      rating,
      comment
    }).then((response) => {
      dispatch(ActionCreator[Action.SET_REVIEWS](response.data));
      dispatch(ActionCreator[Action.SET_COMMENT_SENDING](false));
    }).catch(() => {
      dispatch(ActionCreator[Action.SET_COMMENT_SENDING](false));
      alert(`Something went wrong :(`);
    });
  },

  sortOffers: (sortOptions) => (dispatch, _getState) => {
    const {data: {offers}} = _getState();
    dispatch(ActionCreator[Action.SORT_OPTIONS](sortOptions));
    dispatch(ActionCreator[Action.SET_OFFERS](sortOffers(offers, sortOptions)));
  },

  postFavorite: (propertyId, isFavorite) => (dispatch, _getState, api) => {
    return api.post(`/favorite/${propertyId}/${isFavorite ? 1 : 0}`)
      .then((response) => {
        const {data: newPropertyData = null} = response || {};
        if (newPropertyData) {
          const {
            data: {
              offers = [],
              cities
            }
          } = _getState();
          const index = getOfferIndexById(offers, propertyId);
          if (index >= 0) {
            const newProperty = serverOfferToOffer(newPropertyData, cities);
            offers.splice(index, 1, newProperty);
          }
          dispatch(ActionCreator[Action.SET_OFFERS](offers));
        }
      });
  }
};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case Action.SET_CITIES:
      return Object.assign({}, state, {
        cities: action.payload
      });
    case Action.SET_OFFERS:
      return Object.assign({}, state, {
        offers: action.payload
      });
    case Action.SET_REVIEWS:
      return Object.assign({}, state, {
        reviews: action.payload
      });
    case Action.SORT_OPTIONS:
      return Object.assign({}, state, {
        sort: action.payload
      });
    case Action.SET_FAVORITES:
      return Object.assign({}, state, {
        favorites: action.payload
      });
    case Action.SET_COMMENT_SENDING:
      return Object.assign({}, state, {
        isSendingComment: action.payload
      });
  }
  return state;
};
