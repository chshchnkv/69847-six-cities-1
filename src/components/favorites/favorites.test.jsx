import React from "react";
import {Router} from "react-router-dom";
import renderer from "react-test-renderer";
import Favorites from "./favorites";
import history from "../../history";
import {AccommodationType} from "../../data";
import {Cities} from "../../mocks/cities";

const favoritesMock = [
  {
    cityId: 1,
    offers: [
      {
        id: 2,
        title: `Wood and stone place`,
        src: `img/apartment-02.jpg`,
        isPremium: true,
        isFavorite: true,
        price: 60,
        type: AccommodationType.HOUSE,
        rating: 2.8,
        city: 1,
        location: {
          longitude: 52.369553943508,
          latitude: 4.85309666406198
        },
      },
      {
        id: 3,
        title: `Canal View Prinsengracht`,
        src: `img/apartment-03.jpg`,
        isPremium: false,
        isFavorite: true,
        price: 30,
        type: AccommodationType.HOTEL,
        rating: 4.2,
        city: 1,
        location: {
          longitude: 52.3909553943508,
          latitude: 4.929309666406198
        },
      }
    ]
  },
  {
    cityId: 2,
    offers: [
      {
        id: 1,
        title: `Beautiful & luxurious apartment at great location`,
        src: `img/apartment-01.jpg`,
        isPremium: false,
        isFavorite: true,
        price: 120,
        type: AccommodationType.APARTMENT,
        rating: 3.2,
        city: 2,
        location: {
          longitude: 52.3909553943508,
          latitude: 4.85309666406198
        },
      },
    ]
  }
];

it(`Favorites renders correctly`, () => {
  const onFavoriteClick = jest.fn();
  const onRequestFavorites = jest.fn();
  const tree = renderer.create(
      <Router history={history}>
        <Favorites favorites={favoritesMock} cities={Cities} onFavoriteClick={onFavoriteClick} onRequestFavorites={onRequestFavorites} />
      </Router>
  );
  expect(tree).toMatchSnapshot();
});
