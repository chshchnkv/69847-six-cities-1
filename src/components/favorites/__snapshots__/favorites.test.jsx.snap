// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites renders correctly 1`] = `
Array [
  <main
    className="page__main page__main--favorites"
  >
    <div
      className="page__favorites-container container"
    >
      <section
        className="favorites"
      >
        <h1
          className="favorites__title"
        >
          Saved listing
        </h1>
        <ul
          className="favorites__list"
        >
          <li
            className="favorites__locations-items"
          >
            <div
              className="favorites__locations locations locations--current"
            >
              <div
                className="locations__item"
              >
                <a
                  className="locations__item-link"
                  href="#"
                >
                  <span>
                    Amsterdam
                  </span>
                </a>
              </div>
            </div>
            <div
              className="favorites__places"
            >
              <article
                className="favorites__card place-card"
              >
                <div
                  className="favorites__image-wrapper place-card__image-wrapper"
                >
                  <a
                    href="#"
                    onClick={[Function]}
                  >
                    <img
                      alt="Place image"
                      className="place-card__image"
                      height="200"
                      src=""
                      width="260"
                    />
                  </a>
                </div>
                <div
                  className="favorites__card-info place-card__info"
                >
                  <div
                    className="place-card__price-wrapper"
                  >
                    <div
                      className="place-card__price"
                    >
                      <b
                        className="place-card__price-value"
                      >
                        €
                        60
                      </b>
                      <span
                        className="place-card__price-text"
                      >
                        / 
                        night
                      </span>
                    </div>
                    <button
                      className="place-card__bookmark-button button "
                      onClick={[Function]}
                      type="button"
                    >
                      <svg
                        className="place-card__bookmark-icon"
                        height="19"
                        width="18"
                      >
                        <use
                          xlinkHref="#icon-bookmark"
                        />
                      </svg>
                      <span
                        className="visually-hidden"
                      >
                        To bookmarks
                      </span>
                    </button>
                  </div>
                  <div
                    className="place-card__rating rating"
                  >
                    <div
                      className="place-card__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "60%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                      </span>
                    </div>
                  </div>
                  <h2
                    className="place-card__name"
                  >
                    <a
                      href="/offer/2"
                      onClick={[Function]}
                    >
                      Wood and stone place
                    </a>
                  </h2>
                  <p
                    className="place-card__type"
                  >
                    house
                  </p>
                </div>
              </article>
              <article
                className="favorites__card place-card"
              >
                <div
                  className="favorites__image-wrapper place-card__image-wrapper"
                >
                  <a
                    href="#"
                    onClick={[Function]}
                  >
                    <img
                      alt="Place image"
                      className="place-card__image"
                      height="200"
                      src=""
                      width="260"
                    />
                  </a>
                </div>
                <div
                  className="favorites__card-info place-card__info"
                >
                  <div
                    className="place-card__price-wrapper"
                  >
                    <div
                      className="place-card__price"
                    >
                      <b
                        className="place-card__price-value"
                      >
                        €
                        30
                      </b>
                      <span
                        className="place-card__price-text"
                      >
                        / 
                        night
                      </span>
                    </div>
                    <button
                      className="place-card__bookmark-button button "
                      onClick={[Function]}
                      type="button"
                    >
                      <svg
                        className="place-card__bookmark-icon"
                        height="19"
                        width="18"
                      >
                        <use
                          xlinkHref="#icon-bookmark"
                        />
                      </svg>
                      <span
                        className="visually-hidden"
                      >
                        To bookmarks
                      </span>
                    </button>
                  </div>
                  <div
                    className="place-card__rating rating"
                  >
                    <div
                      className="place-card__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "80%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                      </span>
                    </div>
                  </div>
                  <h2
                    className="place-card__name"
                  >
                    <a
                      href="/offer/3"
                      onClick={[Function]}
                    >
                      Canal View Prinsengracht
                    </a>
                  </h2>
                  <p
                    className="place-card__type"
                  >
                    hotel
                  </p>
                </div>
              </article>
            </div>
          </li>
          <li
            className="favorites__locations-items"
          >
            <div
              className="favorites__locations locations locations--current"
            >
              <div
                className="locations__item"
              >
                <a
                  className="locations__item-link"
                  href="#"
                >
                  <span>
                    Paris
                  </span>
                </a>
              </div>
            </div>
            <div
              className="favorites__places"
            >
              <article
                className="favorites__card place-card"
              >
                <div
                  className="favorites__image-wrapper place-card__image-wrapper"
                >
                  <a
                    href="#"
                    onClick={[Function]}
                  >
                    <img
                      alt="Place image"
                      className="place-card__image"
                      height="200"
                      src=""
                      width="260"
                    />
                  </a>
                </div>
                <div
                  className="favorites__card-info place-card__info"
                >
                  <div
                    className="place-card__price-wrapper"
                  >
                    <div
                      className="place-card__price"
                    >
                      <b
                        className="place-card__price-value"
                      >
                        €
                        120
                      </b>
                      <span
                        className="place-card__price-text"
                      >
                        / 
                        night
                      </span>
                    </div>
                    <button
                      className="place-card__bookmark-button button "
                      onClick={[Function]}
                      type="button"
                    >
                      <svg
                        className="place-card__bookmark-icon"
                        height="19"
                        width="18"
                      >
                        <use
                          xlinkHref="#icon-bookmark"
                        />
                      </svg>
                      <span
                        className="visually-hidden"
                      >
                        To bookmarks
                      </span>
                    </button>
                  </div>
                  <div
                    className="place-card__rating rating"
                  >
                    <div
                      className="place-card__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "60%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                      </span>
                    </div>
                  </div>
                  <h2
                    className="place-card__name"
                  >
                    <a
                      href="/offer/1"
                      onClick={[Function]}
                    >
                      Beautiful & luxurious apartment at great location
                    </a>
                  </h2>
                  <p
                    className="place-card__type"
                  >
                    apartment
                  </p>
                </div>
              </article>
            </div>
          </li>
        </ul>
      </section>
    </div>
  </main>,
  <footer
    className="footer container"
  >
    <a
      className="footer__logo-link"
      href="/"
      onClick={[Function]}
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="img/logo.svg"
        width="64"
      />
    </a>
  </footer>,
]
`;
